zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/client/SharedZkClient.java

```diff
@@ -37,79 +51,549 @@
+          "The given path does not map to the ZK realm for this DedicatedZkClient! Path: " + path
+              + " ZK realm sharding key: " + _zkRealmShardingKey);
+    }
+    return _innerSharedZkClient.getCreationTime(path);
   }
 
   @Override
-  public IZkConnection getConnection() {
-    if (isClosed()) {
-      return IDLE_CONNECTION;
+  public List<OpResult> multi(Iterable<Op> ops) {
+    return _innerSharedZkClient.multi(ops);
+  }
+
+  @Override
+  public boolean waitUntilConnected(long time, TimeUnit timeUnit) {
+    return _innerSharedZkClient.waitUntilConnected(time, timeUnit);
+  }
+
+  @Override
+  public String getServers() {
+    return _innerSharedZkClient.getServers();
+  }
+
+  @Override
+  public long getSessionId() {
+    return _innerSharedZkClient.getSessionId();
+  }
+
+  @Override
+  public void close() {
+    _innerSharedZkClient.close();
+  }
+
+  @Override
+  public boolean isClosed() {
+    return _innerSharedZkClient.isClosed();
+  }
+
+  @Override
+  public byte[] serialize(Object data, String path) {
+    if (!checkIfPathBelongsToZkRealm(path)) {
+      throw new IllegalArgumentException(
+          "The given path does not map to the ZK realm for this DedicatedZkClient! Path: " + path
+              + " ZK realm sharding key: " + _zkRealmShardingKey);
     }
-    return super.getConnection();
+    return _innerSharedZkClient.serialize(data, path);
   }
 
-  /**
-   * Since ZkConnection session is shared in this RealmAwareZkClient, do not create ephemeral node using a SharedZKClient.
-   */
   @Override
-  public String create(final String path, Object datat, final List<ACL> acl,
-      final CreateMode mode) {
-    if (mode.isEphemeral()) {
-      throw new UnsupportedOperationException(
-          "Create ephemeral nodes using a " + SharedZkClient.class.getSimpleName()
-              + " is not supported.");
+  public <T> T deserialize(byte[] data, String path) {
+    if (!checkIfPathBelongsToZkRealm(path)) {
+      throw new IllegalArgumentException(
+          "The given path does not map to the ZK realm for this DedicatedZkClient! Path: " + path
+              + " ZK realm sharding key: " + _zkRealmShardingKey);
     }
-    return super.create(path, datat, acl, mode);
+    return _innerSharedZkClient.deserialize(data, path);
   }
 
   @Override
-  protected boolean isManagingZkConnection() {
-    return false;
+  public void setZkSerializer(ZkSerializer zkSerializer) {
+    _innerSharedZkClient.setZkSerializer(zkSerializer);
+  }
+
+  @Override
+  public void setZkSerializer(PathBasedZkSerializer zkSerializer) {
+    _innerSharedZkClient.setZkSerializer(zkSerializer);
+  }
+
+  @Override
+  public PathBasedZkSerializer getZkSerializer() {
+    return _innerSharedZkClient.getZkSerializer();
+  }
+
+  private boolean checkIfPathBelongsToZkRealm(String path) {
+    // TODO: Check if the path's sharding key equals the sharding key
+    // TODO: Implement this with TrieRoutingData
+    return true;
```

Comment:
As seen here: https://github.com/apache/helix/pull/765/files#diff-851efd495ff678457258d26f3c9f854aR459,
 we could enhance this method to cut down on duplication:

```
    try {
      String zkRealmForPath = _metadataStoreRoutingData.getMetadataStoreRealm(path);
      if (!_zkRealmAddress.equals(zkRealmForPath)) {
        throw new IllegalArgumentException("Given path: " + path + "'s ZK realm: " + zkRealmForPath
            + " does not match the ZK realm: " + _zkRealmAddress + " and sharding key: "
            + _zkRealmShardingKey + " for this DedicatedZkClient!");
      }
    } catch (NoSuchElementException e) {
      throw new IllegalArgumentException(
          "Given path: " + path + " does not have a valid sharding key!");
    }
  }
```

Reply:
done.

Reply:
checkIfPathContainsShardingKey, will do.